pkg/util/deployment/deployment.go

```diff
@@ -142,6 +131,138 @@ func GetNewReplicaSetFromList(deployment extensions.Deployment, c clientset.Inte
+	if err != nil {
+		return nil, nil, err
+	}
+	options := api.ListOptions{LabelSelector: selector}
+	rsList, err := getRSList(namespace, options)
+	if err != nil {
+		return nil, nil, err
+	}
+	syncedRSList := []extensions.ReplicaSet{}
+	for _, rs := range rsList {
+		// Add pod-template-hash information if it's not in the RS.
+		// Otherwise, new RS produced by Deployment will overlap we pre-existing ones
+		// that aren't constrained by the pod-template-hash.
+		syncedRS, err := addHashKeyToRSAndPods(deployment, c, rs, getPodList)
+		if err != nil {
+			return nil, nil, err
+		}
+		syncedRSList = append(syncedRSList, *syncedRS)
+	}
+	syncedPodList, err := getPodList(namespace, options)
+	return syncedRSList, syncedPodList, nil
+}
+
+// addHashKeyToRSAndPods adds pod-template-hash information to the given rs, if it's not already there, with the following steps:
+// 1. Add hash label to all pods this rs owns
+// 2. Add hash label to the rs's pod template, the rs's label, and the rs's selector
+// 3. Clean up all pods this rs owns but without the hash label (orphaned pods)
+func addHashKeyToRSAndPods(deployment extensions.Deployment, c clientset.Interface, rs extensions.ReplicaSet, getPodList podListFunc) (*extensions.ReplicaSet, error) {
+	if labelsutil.SelectorHasLabel(rs.Spec.Selector, extensions.DefaultDeploymentUniqueLabelKey) {
+		return &rs, nil
+	}
+	namespace := deployment.Namespace
+	hash := fmt.Sprintf("%d", podutil.GetPodTemplateSpecHash(*rs.Spec.Template))
+	// 1. Update all pods managed by the rs to have the new hash label, so they will be correctly adopted.
+	selector, err := unversioned.LabelSelectorAsSelector(rs.Spec.Selector)
+	if err != nil {
+		return nil, err
+	}
+	options := api.ListOptions{LabelSelector: selector}
+	podList, err := getPodList(namespace, options)
+	if err != nil {
+		return nil, err
+	}
+	for _, pod := range podList.Items {
+		// If the pod already has the new hash label, avoid re-labeling it
+		if len(pod.Labels) > 0 && len(pod.Labels[extensions.DefaultDeploymentUniqueLabelKey]) > 0 {
+			continue
+		}
+		pod.Labels = labelsutil.AddLabel(pod.Labels, extensions.DefaultDeploymentUniqueLabelKey, hash)
+		delay, maxRetries := 3, 3
+		podName := pod.Name
+		for i := 0; i < maxRetries; i++ {
+			_, err = c.Core().Pods(namespace).Update(&pod)
+			if err == nil {
+				break
+			}
+			time.Sleep(time.Second * time.Duration(delay))
+			delay *= delay
+			getPod, err := c.Core().Pods(namespace).Get(podName)
+			if err != nil {
+				return nil, err
+			}
+			pod = *getPod
+		}
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// 2. Update rs label, rs template label, and rs selector to include the new hash label
+	// Copy the old selector, so that we can scrub out any orphaned pods
+	oldSelector := rs.Spec.Selector
+	// Update the selector of the rs so it manages all the pods we updated above
+	updatedRS, err := updateRSWithRetries(c.Extensions().ReplicaSets(namespace), &rs, func(updated *extensions.ReplicaSet) {
```

Comment:
See https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/rolling_updater.go#L642

That adds the new label to the pod template first in order to avoid races with the controller. A new pod without the label could be added between the loop above and this update.